### PR TITLE
[GPU] Use affine.linearize_index (and delinearize_index) where possible

### PR DESCRIPTION
**Pull request by krzysz00**
There have been issues with the composition of affine maps being too general and loosing important information, like the fact that affine_map<(s0 + s1 * 32 + ... - (s0 floorDiv 16) * 16)> really should be affine_map<(s0 mod 16 + s1 * 32 + ...)>, and other issues with the ultimate IR that block low-level arithmetic optimizations.

The affine.delinearize_index operation represents the div/mod chains needed to break a flat index into its component parts. A recently added affine.linearize_index operation is its inverse - combining multiple indices into a flat 1D value.

Another advantage to linearize/delinearize is simpler upstream canonicalizations and lead to more streamlined generated code.

This PR updates the vector distribution code and other GPU-related code that I could find to

1. Use affine.linearize_index to construct flat thread IDs
2. Use affine.delinearize_index in places where there was a floorDiv/mod chain.

Additionally

1. Change the `scf.for` with a thread ID as initial input approach to non-uniform execution for an `scf.if` to better reflect the intended control flow and enable a linearize
2. Plumb the subgroup size through the transfer_read and transfer_write distribution patterns to enable better reasoning about when you do/don't need to take a mod of the lane ID

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp [hunk 1]
```diff
@@ -143,6 +153,22 @@ LogicalResult resolveGPUMappedForallOp(RewriterBase &rewriter,
                              newBlockArgs);
   rewriter.eraseOp(forallTerminator);
   rewriter.eraseOp(forallOp);
+
+  // Step 5. Create the post-loop code that only executes on some workitems.
+  if (hasPostLoopTail) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp:158]: Could we move this into a separate PR?

**Comment by krzysz00** [comment]
Closing in favor of #19122 so I can split the PR
